# Современные методы анализа данных и машинного обучения, БИ

## НИУ ВШЭ, 2024-25 учебный год

### Домашнее задание №1. Библиотеки Numpy & Pandas

Задание выполнил:

    Хайруллин Дамир

### Общая информация

__Дата выдачи:__ 19.01.2025

__Дедлайн:__ 04:00 29.01.2025

### Оценивание и штрафы

Каждая задача первого раздела (Numpy) оценивается в 2 балла.

Оценка за первый раздел вычисляется по следующей формуле:

$$
s_{\text{numpy}} \times 1/2 ,
$$

где $s_{\text{numpy}}$  — количество баллов, которое вы набрали в сумме по задачам первого раздела.

Каждая задача второго раздела (Pandas) оценивается в 3 балла.

Оценка за второй раздел вычисляется по следующей формуле:

$$
s_{\text{pandas}} \times 10/54 ,
$$

где $s_{\text{pandas}}$  — количество баллов, которое вы набрали в сумме по задачам второго раздела.

Общая оценка за ДЗ вычисляется по следующей формуле:

$$
O_{1} \times 3/10 + O_2 \times 7/10 ,
$$

где $O_1, O_2$  — оценки за первый и второй раздел соответственно.

За сдачу задания позже срока на итоговую оценку за задание накладывается штраф в размере 1 **вторичный** балл в день, но  задержка не может быть больше недели.

**Важно:** В этом домашнем задании предусмотрено несколько заданий, за которые можно получить бонусные баллы. Бонусные баллы прибавляются к обычным баллам, и может выйти так, что ваша итоговая оценка станет больше 10. В таком случае именно это и будет являться вашей оценкой. То есть за это ДЗ **можно** получить больше 10

__Внимание!__ Домашнее задание выполняется самостоятельно. Не допускается помощь в решении домашнего задания от однокурсников или третьих лиц. «Похожие» решения считаются плагиатом, и все задействованные студенты — в том числе и те, у кого списали, — не могут получить за него больше 0 баллов. Использование в решении домашнего задания генеративных моделей (ChatGPT и так далее) за рамками справочной и образовательной информации для генерации кода задания — считается плагиатом, и такое домашнее задание оценивается в 0 баллов.

### Формат сдачи

Загрузка файлов с решениями происходит в системе [Anytask](https://anytask.org/). Необходимо загружать файл с расширением .ipynb (питоновский ноутбук)

Инвайт для подключения к странице курса в системе — будет указан для каждой группы в нашем телеграмм-канале позднее.

## Раздел 1. Numpy

### О задании

В данном разделе вы попрактикуетесь в работе с библиотекой numpy!

In [1]:
import numpy as np

Во всех задачах необходимо написать код решения внутри функции и проверить его работу, вызвав функцию для данных из условия.

При решении задач запрещается использовать циклы (`for`, `while`) и оператор `if`.

Везде, где встречаются массивы или матрицы, подразумевается, что это `numpy.array`.

### 1

Напишите функцию, возвращающую округленную взвешенную сумму оценок по данным оценкам и весам. Можете посчитать свою оценку за курс :) В нашем случае вес экзамена равен 0.25, вес домашек - 0.1, вес групповых проектов - 0.2, вес аудиторки - 0.2, вес контрольной 1 - 0.1, вес контрольной 2 - 0.1 и вес проекта - 0.05. Например, если экзамен написан на 6, домашки сданы на 9, проекты закрыты на 8, оценка за аудиторку 7, контрольная 1 написана на 10, контрольная 2 написана на 7 и проект сдан на 10 - итоговая оценка будет хорошо - 8

In [2]:
def result_mark(weights, marks):
    return np.round(np.sum(weights * marks))

In [3]:
weights = np.array([0.25, 0.1, 0.2, 0.2, 0.1, 0.1, 0.05])
marks = np.array([6, 9, 8, 7, 10, 7, 10])
result_mark(weights, marks)

8.0

### 2

Напишите функцию, меняющую каждое третье (начиная с 0) значение массива целых чисел на заданное число. Например, если на вход поступает массив `array([3, 5, 1, 0, -3, 22, 213436])` и число `-111`, то на выходе должен получиться массив `array([-111, 5, 1, -111, -3, 22, -111])`.

In [4]:
def change_array(array, number):
    array[::3] = number
    return array

In [5]:
array = np.array([3, 5, 1, 0, -3, 22, 213436])
number = -111
change_array(array, number)

array([-111,    5,    1, -111,   -3,   22, -111])

### 3

Напишите функцию, выдающую индексы «близких» элементов заданных массивов, а именно тех пар элементов, чей модуль разницы не превосходит заданного значения. Например, если на вход поступают массив `array([1.5, 0.5, 2, -4.1, -3, 6, -1])`, массив `array([1.2, 0.5, 1, -4, 3, 0, -1.2])` и число `0.5`, то на выходе должен получиться массив `array([0, 1, 3, 6])` (_**важно: не `tuple`, а одномерный массив типа `numpy.ndarray` (то есть `.ndim` от него равно 1)!**_).

In [6]:
def find_close(array1, array2, precision):
    result = np.abs(array1-array2) <= precision
    return np.nonzero(result)[0]

In [8]:
array1 = np.array([1.5, 0.5, 2, -4.1, -3, 6, -1])
array2 = np.array([1.2, 0.5, 1, -4, 3, 0, -1.2])
precision = 0.5
find_close(array1, array2, precision)

array([0, 1, 3, 6])

### 4

Напишите функцию, которая составляет блочную матрицу из четырех блоков, где каждый блок - это заданная матрица. Например, если на вход поступает матрица
$$
\begin{pmatrix}
0 & 1 & 2\\
3 & 4 & 5\\
\end{pmatrix},
$$
то ответом будет матрица
$$
\begin{pmatrix}
0 & 1 & 2 & 0 & 1 & 2\\
3 & 4 & 5 & 3 & 4 & 5\\
0 & 1 & 2 & 0 & 1 & 2\\
3 & 4 & 5 & 3 & 4 & 5\\
\end{pmatrix}
$$

In [9]:
def block_matrix(block):
    block = np.concatenate((block,block), axis = 0)
    block = np.concatenate((block,block), axis = 1)
    return block

In [10]:
block = np.array([[0, 1, 2],[3, 4, 5]])
block_matrix(block)

array([[0, 1, 2, 0, 1, 2],
       [3, 4, 5, 3, 4, 5],
       [0, 1, 2, 0, 1, 2],
       [3, 4, 5, 3, 4, 5]])

### 5

Напишите функцию, вычисляющую произведение всех ненулевых диагональных элементов на главной диагонали данной квадратной матрицы. Например, если на вход поступает матрица
$$
\begin{pmatrix}
0 & 1 & 2\\
3 & 4 & 5\\
6 & 7 & 8\\
\end{pmatrix},
$$
то ответом будет 32.

In [11]:
def diag_prod(matrix):
    diag = np.diag(matrix) > 0
    return np.prod(np.diag(matrix)[diag])

In [12]:
matrix = np.array([[0, 1, 2], [3, 4, 5], [6, 7, 8]])
diag_prod(matrix)

32

### 6

Для улучшения качества работы некоторых алгоритмов машинного обучения может быть полезно использовать [нормализацию данных](https://vk.cc/8xmfQk), чтобы привести признаки в выборке к одному масштабу — а именно, из каждого столбца вычесть среднее его значений и поделить на их стандартное отклонение. Напишите функцию, нормализующую входящую матрицу (по столбцам). Например, если на вход подается матрица
$$
\begin{pmatrix}
1 & 4 & 4200\\
0 & 10 & 5000\\
1 & 2 & 1000\\
\end{pmatrix},
$$
то результатом с точностью до сотых будет матрица
$$
\begin{pmatrix}
0.71 & -0.39 & 0.46\\
-1.41 & 1.37 & 0.93\\
0.71 & -0.98 & -1.39\\
\end{pmatrix}
$$

_Подсказка. Учтите, что в вашей матрице не должно получаться никаких nan. Подумайте, в каком случае они могут возникнуть и как обойти эту проблему._


In [13]:
def normalize(matrix):
    means_to_coloums = np.mean(matrix, axis = 0)
    std_to_colums = np.std(matrix, axis = 0)
    std_to_colums[std_to_colums == 0] = 1
    return np.round((matrix - means_to_coloums) / std_to_colums, 2)

In [14]:
matrix = np.array([[1, 4, 4200], [0, 10, 5000], [1, 2, 1000]])
normalize(matrix)

array([[ 0.71, -0.39,  0.46],
       [-1.41,  1.37,  0.93],
       [ 0.71, -0.98, -1.39]])

### 7

Напишите функцию, вычисляющую какую-нибудь первообразную данного полинома (в качестве константы возьмите ваше любимое число). Например, если на вход поступает массив коэффициентов `array([4, 6, 0, 1])`, что соответствует полиному $4x^3 + 6x^2 + 1$, на выходе получается массив коэффициентов `array([1, 2, 0, 1, -2])`, соответствующий полиному $x^4 + 2x^3 + x - 2$.

In [15]:
def antiderivative(coefs):
    return np.polyint(coefs, k=16)

In [16]:
coefs = np.array([4, 6, 0, 1])
antiderivative(coefs)

array([ 1.,  2.,  0.,  1., 16.])

### 8

Напишите функцию, делающую данную [треугольную матрицу](https://ru.wikipedia.org/wiki/%D0%A2%D1%80%D0%B5%D1%83%D0%B3%D0%BE%D0%BB%D1%8C%D0%BD%D0%B0%D1%8F_%D0%BC%D0%B0%D1%82%D1%80%D0%B8%D1%86%D0%B0) симметричной. Например, если на вход поступает матрица
$$
\begin{pmatrix}
1 & 2 & 3 & 4\\
0 & 5 & 6 & 7\\
0 & 0 & 8 & 9\\
0 & 0 & 0 & 10\\
\end{pmatrix},
$$
то на выходе должна быть матрица
$$
\begin{pmatrix}
1 & 2 & 3 & 4\\
2 & 5 & 6 & 7\\
3 & 6 & 8 & 9\\
4 & 7 & 9 & 10\\
\end{pmatrix}.
$$

In [17]:
def make_symmetric(matrix):
    return matrix + matrix.T - np.diag(matrix.diagonal())

In [18]:
matrix = np.array([[1, 2, 3, 4], [0, 5, 6, 7], [0, 0, 8, 9], [0, 0, 0, 10]])
make_symmetric(matrix)

array([[ 1,  2,  3,  4],
       [ 2,  5,  6,  7],
       [ 3,  6,  8,  9],
       [ 4,  7,  9, 10]])

### 9

Напишите функцию, создающую прямоугольную матрицу из m одинаковых строк, заполненных последовательными натуральными числами от a до b включительно в возрастающем порядке. Например, если m = 5, a = 3, b = 10, то на выходе будет матрица
$$
\begin{pmatrix}
3 & 4 & 5 & 6 & 7 & 8 & 9 & 10\\
3 & 4 & 5 & 6 & 7 & 8 & 9 & 10\\
3 & 4 & 5 & 6 & 7 & 8 & 9 & 10\\
3 & 4 & 5 & 6 & 7 & 8 & 9 & 10\\
3 & 4 & 5 & 6 & 7 & 8 & 9 & 10\\
\end{pmatrix}
$$

In [19]:
def construct_matrix(m, a, b):
    ar = np.arange(a,b+1)
    return np.tile(ar, (m,1))

In [20]:
m = 5
a = 3
b = 10
construct_matrix(m, a, b)

array([[ 3,  4,  5,  6,  7,  8,  9, 10],
       [ 3,  4,  5,  6,  7,  8,  9, 10],
       [ 3,  4,  5,  6,  7,  8,  9, 10],
       [ 3,  4,  5,  6,  7,  8,  9, 10],
       [ 3,  4,  5,  6,  7,  8,  9, 10]])

### 10

Напишите функцию, вычисляющую [косинусную близость](https://en.wikipedia.org/wiki/Cosine_similarity) двух векторов. Например, если на вход поступают вектора `array([-2, 1, 0, -5, 4, 3, -3])` и `array([0, 2, -2, 10, 6, 0, 0])`, ответом будет -0.25.

In [21]:
def cosine_similarity(vec1, vec2):
    return np.dot(vec1, vec2)/(np.linalg.norm(vec1) * np.linalg.norm(vec2))

In [22]:
vec1 = np.array([-2, 1, 0, -5, 4, 3, -3])
vec2 = np.array([0, 2, -2, 10, 6, 0, 0])
cosine_similarity(vec1, vec2)

-0.25

## Раздел 2. Pandas

### О задании

В данном разделе вы попрактикуетесь в работе с библиотекой pandas!

In [26]:
import pandas as pd

Обратите внимание, что практически во всех заданиях вам потребуется написать какой-то текст в виде вывода/результата/итога/какого-то резюме. Пожалуйста, не ленитесь, и пишите выводы ***подробно***, существенная часть баллов будет выставляться именно за выводы, так как в анализе данных они имеют большое значение. Таким образом, на любое задание, сформулированное в виде вопроса, отвечайте текстом, подкрепляя ваши слова написанным и выполненным кодом. Текст выводов/комментариев пишите в отдельных ячейках, используя средства текстового редактора Jupyter Notebook/Google Colab, при необходимости используйте средства *выделения*.

При решении задач, когда вам необходимо сделать какой-то фильтр, срез, выборку, или иное действие с данными, запрещается использовать циклы (`for`, `while`) и оператор `if`.


### 1

В этом разделе домашнего задания вы поработаете с данными исследования зависимости средней продолжительности жизни по странам от количества употребляемого в стране алкоголя.

Вашему вниманию представлено два датасета: drinks.csv и lifeexpectancy-verbose.csv

Первая таблица содержит в себе сводные данные по количеству употребляемого алкоголя в каждой стране.

Описание датасета:
- country -	Name of the country (String)
- beer_servings	- Number of beer servings per person in the country (Integer)
- spirit_servings	- Number of spirit servings per person in the country (Integer)
- wine_servings	- Number of wine servings per person in the country(Integer)
- total_litres_of_pure_alcohol -	Total litres of pure alcohol consumed per person in the country (Float)

Вторая таблица содержит в себе информацию о научных статьях и исследованиях о продолжительности жизни (в разных категориях и для разных групп) по той или иной стране.

Откройте в ноутбуке оба файла с данными. Какой размер имеют таблицы? Выведите первые 5 и последние 5 строк для обеих таблиц. Сколько уникальных стран представлено в каждой из таблиц? Сделайте выводы.

In [28]:
drinks = pd.read_csv('drinks.csv')
print('Размер таблицы drinks:', drinks.shape)

Размер таблицы drinks: (193, 6)


In [29]:
drinks.head(5)

index      country  beer_servings  spirit_servings  wine_servings  \
0      0  Afghanistan              0                0              0   
1      1      Albania             89              132             54   
2      2      Algeria             25                0             14   
3      3      Andorra            245              138            312   
4      4       Angola            217               57             45   

   total_litres_of_pure_alcohol  
0                           0.0  
1                           4.9  
2                           0.7  
3                          12.4  
4                           5.9

In [30]:
drinks.tail(5)

index    country  beer_servings  spirit_servings  wine_servings  \
188    188  Venezuela            333              100              3   
189    189    Vietnam            111                2              1   
190    190      Yemen              6                0              0   
191    191     Zambia             32               19              4   
192    192   Zimbabwe             64               18              4   

     total_litres_of_pure_alcohol  
188                           7.7  
189                           2.0  
190                           0.1  
191                           2.5  
192                           4.7

In [31]:
print('Количество стран в датасете drinks:', drinks['country'].nunique())

Количество стран в датасете drinks: 193


In [32]:
life_expect = pd.read_csv('lifeexpectancy-verbose.csv')
print('Размер таблицы lifeexpectancy-verbose:',life_expect.shape)

Размер таблицы lifeexpectancy-verbose: (6408, 17)


In [33]:
life_expect.head(5)

index        GhoCode                                       GhoDisplay  \
0      0  WHOSIS_000001                 Life expectancy at birth (years)   
1      1  WHOSIS_000002  Healthy life expectancy (HALE) at birth (years)   
2      2  WHOSIS_000001                 Life expectancy at birth (years)   
3      3  WHOSIS_000001                 Life expectancy at birth (years)   
4      4  WHOSIS_000015                Life expectancy at age 60 (years)   

  PublishStateCode PublishStateDisplay  YearCode  YearDisplay RegionCode  \
0        PUBLISHED           Published      1990         1990        AMR   
1        PUBLISHED           Published      2012         2012        EUR   
2        PUBLISHED           Published      2000         2000        EMR   
3        PUBLISHED           Published      2000         2000        AFR   
4        PUBLISHED           Published      1990         1990       SEAR   

           RegionDisplay WorldBankIncomeGroupGroupCode  \
0               Americas                        WB_LMI   
1                 Europe                         WB_HI   
2  Eastern Mediterranean                         WB_LI   
3                 Africa                        WB_LMI   
4        South_East Asia                        WB_LMI   

  WorldBankIncomeGroupDisplay CountryCode CountryDisplay SexCode  SexDisplay  \
0         Lower_middle_income         NIC      Nicaragua     MLE        Male   
1                 High_income         IRL        Ireland     MLE        Male   
2                  Low_income         YEM          Yemen    BTSX  Both sexes   
3         Lower_middle_income         NGA        Nigeria    FMLE      Female   
4         Lower_middle_income         THA       Thailand    FMLE      Female   

   DisplayValue  Numeric  
0            68     68.0  
1            69     69.3  
2            61     61.0  
3            48     48.0  
4            19     19.0

In [34]:
life_expect.tail(5)

index        GhoCode                                       GhoDisplay  \
6403   6403  WHOSIS_000001                 Life expectancy at birth (years)   
6404   6404  WHOSIS_000015                Life expectancy at age 60 (years)   
6405   6405  WHOSIS_000001                 Life expectancy at birth (years)   
6406   6406  WHOSIS_000002  Healthy life expectancy (HALE) at birth (years)   
6407   6407  WHOSIS_000001                 Life expectancy at birth (years)   

     PublishStateCode PublishStateDisplay  YearCode  YearDisplay RegionCode  \
6403        PUBLISHED           Published      2000         2000        EMR   
6404        PUBLISHED           Published      1990         1990        WPR   
6405        PUBLISHED           Published      1990         1990        AFR   
6406        PUBLISHED           Published      2012         2012        EUR   
6407        PUBLISHED           Published      2013         2013        EUR   

              RegionDisplay WorldBankIncomeGroupGroupCode  \
6403  Eastern Mediterranean                        WB_LMI   
6404        Western Pacific                        WB_UMI   
6405                 Africa                         WB_LI   
6406                 Europe                        WB_LMI   
6407                 Europe                           NaN   

     WorldBankIncomeGroupDisplay CountryCode CountryDisplay SexCode  \
6403         Lower_middle_income         MAR        Morocco    BTSX   
6404         Upper_middle_income         PLW          Palau    BTSX   
6405                  Low_income         BFA   Burkina Faso    FMLE   
6406         Lower_middle_income         ALB        Albania     MLE   
6407                         NaN         ARM        Armenia    BTSX   

      SexDisplay  DisplayValue   Numeric  
6403  Both sexes            68  68.00000  
6404  Both sexes            16  16.00000  
6405      Female            51  51.00000  
6406        Male            64  63.50000  
6407  Both sexes            71  71.02699

In [35]:
print('Количество стран в датасете lifeexpectancy-verbose.csv:', life_expect['CountryDisplay'].nunique())

Количество стран в датасете lifeexpectancy-verbose.csv: 194


# **Выводы:**

1. Количество стран в датасетах отлично на 1 -  ***drinks.csv*** включает 193 страны, когда как ***lifeexpectancy-verbose.csv*** 194. Это значит, что записи, одной из страны не будут включены в новый датасет, если мы сделаем inner join, так как по ней нет статистики с таблицы drinks.
2. Размерность датасета ***drinks.csv*** (193, 6)
3. Размерность датасета ***lifeexpectancy-verbose.csv*** (6408, 17)








### 2

Для каждого признака в рассматриваемых данных укажите вид этого признака и кратко обоснуйте, почему вид именно такой.

# **Для таблицы drinks:**


1.   **country** - категориальный (название страны, категории без порядка)
2.   **beer_servings** - вещественный (количество пива на человека, числовое значение)
3. **spirit_servings** - вещественный (количество крепкого алко. на человека, также числовое значение)
4. **wine_servings** - вещественный (количество вина, как и предыдущие числовое значение)
5. **total_litres_of_pure_alcohol** - вещественный (суммарное количество алко. на человека в стране, тоже числовое)


# **Для таблицы lifeexpectancy-verbose:**



1.   **GhoCode** - категораильный (уникальный код каждой записи)
2.   **GhoDisplay** - категориальный (название вида записи, например "Ожидаемая продолжительность жизни при рождении", без порядка)
3. **PublishStateCode** - бинарный (состояние "опубликовно/не опубликовано" подразумевается что признак может принимать лишь 2 значения, но в данном датасете лишь опубликованные записи)
4. **PublishStateDisplay** - бинарный (дублирует признак выше, в самом датасете также лишь статус опубликованно)
5. **YearCode** - вещественный (код года публикации, совпадает с реальным годом)
6. **YearDisplay** - вещественный (дублирует признак выше, выводит год)
7. **RegionCode** - категориальный (буквенный код региона)
8. **RegionDisplay** - категориальный (полное название региона)
9. **WorldBankIncomeGroupGroupCode** - порядковый (кодовое название группы доходов)
10. **WorldBankIncomeGroupDisplay** - порядковый (полное название групп доходов, так же есть порядок (lower_middle, high и тд.))
11. **CountryCode** - категориальный (буквенный код страны)
12. **CountryDisplay** - категориальный (полное название страны)
13. **SexCode** - категориальный (кодове название полов, на основе которых создана запись)
14. **SexDisplay** - категориальный (полное название полов (мужской, женский, оба))
15. **DisplayValue** - вещетсвенный (числовой показатель **GhoDisplay** в формате int)
16. **Numeric** - вещественный (дублирует показатель сверху, но в формате float)





### 3

Внимательно изучите признаки обеих таблиц. Как вы считаете, какая из таблиц является главной (а какая сопутствующей)? Почему вы так думаете?

Какой признак может представлять существенный интерес для исследования и являться целевой переменной в рамках представленного датасета? Ответ обоснуйте.

***Бонус (1 бон. балл):*** Попробуйте описать задачу машинного обучения, которую вы могли бы решать для данного датасета с учетом выбранной вами раннее целевой переменной и проблематики исследования. Подробнее про постановку и формализацию задач машинного обучения, а также про различные типы таких задач — вы можете прочитать, например, вот в этой [статье](https://habr.com/ru/company/simbirsoft/blog/565396/) (в начале)



1.   Главная - lifeexpectancy-verbose, сопутутствующая - drinks. Первая содержит в себе комплексную статистику по длине жизни взависимости от года, пола, места и соц. группы. Таблицу drinks тут актуальнее использовать в качестве вспомогательной, для нахождении взаимосвязей.
2.   Это признаки **DisplayValue/Numeric**. Наиболее вероятный вариант, скорее мы хотим научиться предсказывать длину жизни человека от различных обстоятельств, окружающих его.
3. Нам нужен числовой прогноз, следовательно в машинном обучении это **задача регрессии**. Если попытаться описать ее  - "Спрогнозировать ожидаемую продолжительность жизни на основе признаков, связанных с соц. положением, географическим расположением, потреблением алкоголя и пола".




### 4

Какова доля статей о продолжительности жизни в Африке? Какова доля статей о продолжительности жизни мужчин и доля статей о продолжительности жизни женщин среди всех статей по Африке?

***Бонус (2 бон. балла):*** Сравните эти доли с аналогичными долями по другим регионам. Как вы можете объяснить возникающие сходства и различия? Прокомментируйте полученные результаты.

In [36]:
africa_stat = life_expect[life_expect['RegionDisplay'] == 'Africa']
africa_stat_proportion = len(africa_stat)/len(life_expect)
men_in_africa = len(africa_stat[africa_stat['SexDisplay'] == 'Male']) / len(africa_stat)
women_in_africa = len(africa_stat[africa_stat['SexDisplay'] == 'Female']) / len(africa_stat)
print(
    f"Доля статей о продолжительности жизни в Африке: {africa_stat_proportion:.1%}\n"
    f"Доля статей о продолжительности жизни мужчин: {men_in_africa :.1%}\n"
    f"Доля статей о продолжительности жизни женщин: {women_in_africa:.1%}"
)

Доля статей о продолжительности жизни в Африке: 23.8%
Доля статей о продолжительности жизни мужчин: 33.3%
Доля статей о продолжительности жизни женщин: 33.3%


In [37]:
regions = life_expect.groupby('RegionDisplay')
region_proportion = regions.size()/ len(life_expect)
men_proport = regions.apply(lambda i: len(i[i['SexDisplay'] == 'Male']) / len(i))
women_proport = regions.apply(lambda i: len(i[i['SexDisplay'] == 'Female']) / len(i))
result = pd.DataFrame({
    'Доля статей о регионе': region_proportion,
    'Доля мужчин': men_proport,
    'Доля женщин': women_proport})

result

<ipython-input-37-f822c65ef127>:3: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  men_proport = regions.apply(lambda i: len(i[i['SexDisplay'] == 'Male']) / len(i))
<ipython-input-37-f822c65ef127>:4: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  women_proport = regions.apply(lambda i: len(i[i['SexDisplay'] == 'Female']) / len(i))


Доля статей о регионе  Доля мужчин  Доля женщин
RegionDisplay                                                         
Africa                              0.237828     0.333333     0.333333
Americas                            0.180243     0.333333     0.333333
Eastern Mediterranean               0.113296     0.333333     0.333333
Europe                              0.272940     0.333333     0.333333
South_East Asia                     0.056648     0.333333     0.333333
Western Pacific                     0.139045     0.333333     0.333333

**# Выводы:**



1.   Наибольшие доли статей приходится на Африку (0.237) и Европу (0.27). Справдливо предположить, что исследователи старались набрать данные с контрастирующих регионов (в Африке дела с здравоохранением хуже, чем в Европе). Америка заняла 3 место с долей 0.18, также развитый регион с высокой продолжительностью жизни. Регион Eastern Mediterranean и Western Pacific с долями 0.11 и 0.13, соответсвенно. По всей видимости доли таковы, потому что результаты схожи с развитыми странами, либо была проблема с сбором данных. Эти же причины применитльный к South_East Asia.  
2.   Говоря о гендерных долях, мы видим стабильное разделение по 0.33 для каждого. На мой взгляд это сделано для сбалансированности выборки, во избежании перевесов.




### 5

Сколько всего существует видов статей с точки зрения оцениваемого параметра продолжительности жизни? Какой из этих видов статей самый популярный (написано наибольшее количество научных работ)? В рамках найденного вами самого популярного вида статей проведите сравнение, относительно какой из половых групп: мужчины, женщины или же оба пола — статей в нем было больше всего?

In [38]:
print('Видов статей :', life_expect['GhoDisplay'].nunique())

Видов статей : 3


In [39]:
art_types = life_expect['GhoDisplay'].value_counts().reset_index()
art_types.columns = ['Вид статьи', 'Количество записей']
art_types

Вид статьи  Количество записей
0                 Life expectancy at birth (years)                2328
1                Life expectancy at age 60 (years)                2328
2  Healthy life expectancy (HALE) at birth (years)                1752

У Life expectancy at birth (years) и Life expectancy at age 60 (years) одинаково по 2328 записей

In [40]:
popular_variant = life_expect[life_expect['GhoDisplay'] == 'Life expectancy at birth (years)']
gender_proportion = popular_variant['SexDisplay'].value_counts()
gender_proportion

SexDisplay
Male          776
Both sexes    776
Female        776
Name: count, dtype: int64

Рассмотрим вид Life expectancy at birth (years). Наблюдаем, что распредление равномерное, нет ни малейшего перепада в одну из сторон.

In [41]:
popular_variant_2 = life_expect[life_expect['GhoDisplay'] == 'Life expectancy at age 60 (years)']
gender_proportion = popular_variant_2['SexDisplay'].value_counts()
gender_proportion

SexDisplay
Female        776
Both sexes    776
Male          776
Name: count, dtype: int64

У Life expectancy at age 60 (years) ровно так же.

### 6

Выведите минимальную, среднюю, медианную и максимальную оцененную продолжительность жизни людей от рождения (life expectancy at birth). Выберите любые три страны на ваш вкус и проделайте то же самое только для статей по этим странам. Что можно сказать на основе результатов, которые вы получили?

In [42]:
min_life = popular_variant['Numeric'].min()
max_life = popular_variant['Numeric'].max()
mean_life = popular_variant['Numeric'].mean()
median_life = popular_variant['Numeric'].median()
print(
    f"Минимальная продолжительность жизни людей от рождения: {min_life:.1f}\n"
    f"Средняя продолжительность жизни людей от рождения: {mean_life:.1f}\n"
    f"Медианная продолжительность жизни людей от рождения: {median_life:.1f}\n"
    f"Максимальная продолжительность жизни людей от рождения: {max_life:.1f}"
)

Минимальная продолжительность жизни людей от рождения: 38.0
Средняя продолжительность жизни людей от рождения: 68.4
Медианная продолжительность жизни людей от рождения: 70.9
Максимальная продолжительность жизни людей от рождения: 87.2


In [43]:
countries = ['Russian Federation', 'Japan', 'Nigeria']

data_about_countries = life_expect[(life_expect['GhoDisplay'] == 'Life expectancy at birth (years)') &
(life_expect['CountryDisplay'].isin(countries))]
stata = data_about_countries.groupby('CountryDisplay')['Numeric'].agg(
    Min='min',
    Mean='mean',
    Median='median',
    Max='max')

stata

Min       Mean  Median       Max
CountryDisplay                                       
Japan               76.0  81.925610    81.5  87.15457
Nigeria             45.0  50.473338    50.5  55.40039
Russian Federation  59.0  68.078784    69.0  75.22597

Выбрана Япония, для которой общеизвестным фактом является высокая продолжительнсть жизни, Нигеря, африканская страна с проблемами в здравоохранении, и Россия, для сравнения относительно худшего и лучшего. У трех стран минимальная продолжительность жизни превышает мин. продолжительность жизни общую, что является хорошим фактом. Но со средней и медианной не все так просто. Стаистика по средней продолжительности по России соответсвует общей, а медианная чуть ниже по миру. У Нигерии все печальнее, средняя отстает на примерно 18 лет, а медианная на почти целых 20 лет, это же касается максимальной длины - она всего 55 лет, отстает от мирового уровня на почти 35 лет. У Японии, как и ожидалось, все намного позитивнее, средняя и мадеианная значительно выше общемирового, а максимум по общей статистике брался скорее всего по ней, страна - эталон.

Вывод по России: длина жизни по рассмотренным метрикам ощутимо выше, чем у страны с проблемами в здравоохранении, но еще далека от страны - эталона.

### 7

Сформируйте результирующий датафрейм на основе проделанных вами действий в предыдущем номере. У датафрейма должно быть 4 столбца, в соответствии с заданными агреггирующими функциями: минимум, среднее, медиана, максимум, — а также 4 строки: статистика по всему датасету; статистика по трем выбранным вами странам. Сохраните программно созданный вами датафрейм в виде csv файла.

*Примечание. Датафрейм необходимо создавать автоматизированным образом. Не допускается ручной перенос значений из полученных в предыдущем задании ячеек в пустой датафрейм и/или прописывание конкретных значений (чисел) в коде программы. Представьте, что у нас не 3 страны, а 100 и не 4 агреггирующих метрики, а 100!*

In [44]:
world_stat = life_expect[life_expect['GhoDisplay'] == 'Life expectancy at birth (years)']['Numeric'].agg(
    Min='min',
    Mean='mean',
    Median='median',
    Max='max').to_frame().T
world_stat['CountryDisplay'] = 'Global'

country_stats = data_about_countries.groupby('CountryDisplay')['Numeric'].agg(
    Min='min',
    Mean='mean',
    Median='median',
    Max='max').reset_index()

final_df = pd.concat([world_stat, country_stats], ignore_index=True)
final_df = final_df[['CountryDisplay', 'Min', 'Mean', 'Median', 'Max']]
final_df.to_csv('df', index=False)
final_df

CountryDisplay   Min       Mean     Median       Max
0              Global  38.0  68.441379  70.857895  87.15457
1               Japan  76.0  81.925610  81.500000  87.15457
2             Nigeria  45.0  50.473338  50.500000  55.40039
3  Russian Federation  59.0  68.078784  69.000000  75.22597

### 8

Есть ли в наших датасетах пропуски? Если да, то в каких столбцах? Сколько пропусков в каждом из них?

Подумайте, как стоит избавиться от пропусков в каждом из столбцов, и сделайте это. Аргументируйте ваши действия.

Проверьте, что пропусков нигде не осталось.

*Примечание. Не забудьте сохранить все изменения! В последующих заданиях вы должны работать с уже обновленными датасетами.*

In [45]:
drinks.isnull().sum()

index                           0
country                         0
beer_servings                   0
spirit_servings                 0
wine_servings                   0
total_litres_of_pure_alcohol    0
dtype: int64

In [46]:
life_expect.isnull().sum()

index                               0
GhoCode                             0
GhoDisplay                          0
PublishStateCode                    0
PublishStateDisplay                 0
YearCode                            0
YearDisplay                         0
RegionCode                          0
RegionDisplay                       0
WorldBankIncomeGroupGroupCode    1164
WorldBankIncomeGroupDisplay      1164
CountryCode                         0
CountryDisplay                      0
SexCode                             0
SexDisplay                          0
DisplayValue                        0
Numeric                             0
dtype: int64

Как видим пропуски есть только в датасете life_expectancy в столбцах с социальной принадлежностью. На мой взгляд это важный признак в нашем исследовании, с ним у таргета потенциально может быть сильная корреляция. Поэтому заполнять его часто встречающимся значением было бы опрометчиво, это может навредить точности модели. Исходя из этого, я предлагаю их удалить, таких записей 1164 из 6407, что не критично на первый взгляд.

In [47]:
life_expect = life_expect.dropna(subset=['WorldBankIncomeGroupGroupCode', 'WorldBankIncomeGroupDisplay'])
life_expect.isnull().sum()

index                            0
GhoCode                          0
GhoDisplay                       0
PublishStateCode                 0
PublishStateDisplay              0
YearCode                         0
YearDisplay                      0
RegionCode                       0
RegionDisplay                    0
WorldBankIncomeGroupGroupCode    0
WorldBankIncomeGroupDisplay      0
CountryCode                      0
CountryDisplay                   0
SexCode                          0
SexDisplay                       0
DisplayValue                     0
Numeric                          0
dtype: int64

Пропуски успешно устранены

### 9

Все ли признаки несут в себе полезную информацию? Почему вы так считаете? Избавьтесь от ненужных столбцов. Свой выбор аргументируйте

*Примечание. Не забудьте сохранить все изменения! В последующих заданиях вы должны работать с уже обновленными датасетами.*

In [48]:
columns_to_drop = [
    'index', # есть автоматическая индексация
    'GhoCode', # не несет полезной информации
    'PublishStateCode',  # лишь один статус, не несет смысл в анализе
    'PublishStateDisplay', # дублирует признак выше
    'YearCode', # дублирует YearDisplay
    'RegionCode', #  кодовые сокращения
    'WorldBankIncomeGroupGroupCode', # кодовые сокращения
    'CountryCode', # кодовые сокращения
    'SexCode', # кодовые сокращения
    'DisplayValue' # Numeric в int, может испортить точность модели
]
life_expect = life_expect.drop(columns=columns_to_drop)
life_expect

GhoDisplay  YearDisplay  \
0                    Life expectancy at birth (years)         1990   
1     Healthy life expectancy (HALE) at birth (years)         2012   
2                    Life expectancy at birth (years)         2000   
3                    Life expectancy at birth (years)         2000   
4                   Life expectancy at age 60 (years)         1990   
...                                               ...          ...   
6401                 Life expectancy at birth (years)         2000   
6403                 Life expectancy at birth (years)         2000   
6404                Life expectancy at age 60 (years)         1990   
6405                 Life expectancy at birth (years)         1990   
6406  Healthy life expectancy (HALE) at birth (years)         2012   

              RegionDisplay WorldBankIncomeGroupDisplay CountryDisplay  \
0                  Americas         Lower_middle_income      Nicaragua   
1                    Europe                 High_income        Ireland   
2     Eastern Mediterranean                  Low_income          Yemen   
3                    Africa         Lower_middle_income        Nigeria   
4           South_East Asia         Lower_middle_income       Thailand   
...                     ...                         ...            ...   
6401               Americas         Lower_middle_income       Honduras   
6403  Eastern Mediterranean         Lower_middle_income        Morocco   
6404        Western Pacific         Upper_middle_income          Palau   
6405                 Africa                  Low_income   Burkina Faso   
6406                 Europe         Lower_middle_income        Albania   

      SexDisplay  Numeric  
0           Male     68.0  
1           Male     69.3  
2     Both sexes     61.0  
3         Female     48.0  
4         Female     19.0  
...          ...      ...  
6401        Male     69.0  
6403  Both sexes     68.0  
6404  Both sexes     16.0  
6405      Female     51.0  
6406        Male     63.5  

[5244 rows x 7 columns]

### 10

Посчитайте, насколько сильно коррелируют друг с другом потребления разных видов алкоголя (beer_servings,	spirit_servings и	wine_servings). Создайте датафрейм с попарными корреляциями этих столбцов друг с другом. Какие выводы можно сделать?

In [49]:
corr_matrix = drinks[['beer_servings', 'spirit_servings', 'wine_servings']].corr()
corr_matrix

beer_servings  spirit_servings  wine_servings
beer_servings         1.000000         0.458819       0.527172
spirit_servings       0.458819         1.000000       0.194797
wine_servings         0.527172         0.194797       1.000000

Наивысшая корреляцию наблюдается у вина с пивом, она достигает 0.527, можно выдвинуть предположение, что в некоторых странах, где любят потреблять пиво, не редкость, что также преобладает вино, но связь не настолько сильна, чтобы говорить о явной зависмости. То же самое можно сказать и о корреляции между пивом и крепким. Что же касается связи между крепким и вином, то она совсем слабая и достигает значения около 0.2

### 11

Каким образом следует осуществить соединение двух имеющихся у нас таблиц? Почему? Соедините таблицы!

Получилось ли у вас задействовать все строки, и если нет, то почему? Что делать с теми строками, которые не удается задействовать, если таковые есть? Ответ обоснуйте.

*Подсказка. В ответе на первый вопрос задания, а также в обосновании способа соединения таблиц вам может помочь задание 3 данного раздела ДЗ и соответствующий семинарский ноутбук!*

*Примечание. Не забудьте сохранить все изменения! В последующих заданиях вы должны работать с одним объединенным датасетом.*

Соединение датасетов само собой необходимо делать по совпадаещему признаку, в нашем случае это страны. Проверим совпадают ли их названия точь в точь

In [50]:
set(drinks['country'].unique()) - set(life_expect['CountryDisplay'].unique())

{'Antigua & Barbuda',
 'Bolivia',
 'Bosnia-Herzegovina',
 'Brunei',
 "Cote d'Ivoire",
 'DR Congo',
 'Guinea-Bissau',
 'Iran',
 'Laos',
 'Macedonia',
 'Micronesia',
 'Moldova',
 'North Korea',
 'Sao Tome & Principe',
 'South Korea',
 'St. Kitts & Nevis',
 'St. Lucia',
 'St. Vincent & the Grenadines',
 'Syria',
 'Tanzania',
 'Timor-Leste',
 'Trinidad & Tobago',
 'USA',
 'United Kingdom',
 'Venezuela',
 'Vietnam'}

In [51]:
set(life_expect['CountryDisplay'].unique()) - set(drinks['country'].unique())

{'Antigua and Barbuda',
 'Bolivia (Plurinational State of)',
 'Bosnia and Herzegovina',
 'Brunei Darussalam',
 "Côte d'Ivoire",
 "Democratic People's Republic of Korea",
 'Democratic Republic of the Congo',
 'Guinea_Bissau',
 'Iran (Islamic Republic of)',
 "Lao People's Democratic Republic",
 'Micronesia (Federated States of)',
 'Republic of Korea',
 'Republic of Moldova',
 'Saint Kitts and Nevis',
 'Saint Lucia',
 'Saint Vincent and the Grenadines',
 'Sao Tome and Principe',
 'South Sudan',
 'Syrian Arab Republic',
 'The former Yugoslav republic of Macedonia',
 'Timor_Leste',
 'Trinidad and Tobago',
 'United Kingdom of Great Britain and Northern Ireland',
 'United Republic of Tanzania',
 'United States of America',
 'Venezuela (Bolivarian Republic of)',
 'Viet Nam'}

Как видим у нас проблемы, названия целого ряда стран не совпадают точь в точь. Одно из решении, это привести их к одному названию и только потом произвести слияние. Этим и займемся, нам поможет маппинг.

In [54]:
mapping = {
    'Iran (Islamic Republic of)': 'Iran',
    'Republic of Korea': 'South Korea',
    'Syrian Arab Republic': 'Syria',
    'Lao People\'s Democratic Republic': 'Laos',
    'Democratic People\'s Republic of Korea': 'North Korea',
    'United Republic of Tanzania': 'Tanzania',
    'Republic of Moldova': 'Moldova',
    'Viet Nam': 'Vietnam',
    'Venezuela (Bolivarian Republic of)': 'Venezuela',
    'United Kingdom of Great Britain and Northern Ireland': 'United Kingdom',
    'Micronesia (Federated States of)': 'Micronesia',
    'Trinidad and Tobago': 'Trinidad & Tobago',
    'Democratic Republic of the Congo': 'DR Congo',
    'Brunei Darussalam': 'Brunei',
    'Saint Vincent and the Grenadines': 'St. Vincent & the Grenadines',
    'Timor_Leste': 'Timor-Leste',
    'United States of America': 'USA',
    'Côte d\'Ivoire': 'Cote d\'Ivoire',
    'Bolivia (Plurinational State of)': 'Bolivia',
    'The former Yugoslav republic of Macedonia': 'Macedonia',
    'Saint Lucia': 'St. Lucia',
    'Saint Kitts and Nevis': 'St. Kitts & Nevis',
    'Antigua and Barbuda': 'Antigua & Barbuda',
    'Bosnia and Herzegovina': 'Bosnia-Herzegovina',
    'Guinea-Bissau': 'Guinea_Bissau',
    'Sao Tome and Principe': 'Sao Tome & Principe'
}

life_expect['CountryDisplay'] = life_expect['CountryDisplay'].replace(mapping)


df_mrg = pd.merge(drinks.rename(columns={'country': 'CountryDisplay'}),
                     life_expect,
                     on='CountryDisplay',
                     how='outer')
df_mrg = df_mrg.drop(columns='index')
df_mrg

CountryDisplay  beer_servings  spirit_servings  wine_servings  \
0       Afghanistan            0.0              0.0            0.0   
1       Afghanistan            0.0              0.0            0.0   
2       Afghanistan            0.0              0.0            0.0   
3       Afghanistan            0.0              0.0            0.0   
4       Afghanistan            0.0              0.0            0.0   
...             ...            ...              ...            ...   
5240       Zimbabwe           64.0             18.0            4.0   
5241       Zimbabwe           64.0             18.0            4.0   
5242       Zimbabwe           64.0             18.0            4.0   
5243       Zimbabwe           64.0             18.0            4.0   
5244       Zimbabwe           64.0             18.0            4.0   

      total_litres_of_pure_alcohol  \
0                              0.0   
1                              0.0   
2                              0.0   
3                              0.0   
4                              0.0   
...                            ...   
5240                           4.7   
5241                           4.7   
5242                           4.7   
5243                           4.7   
5244                           4.7   

                                           GhoDisplay  YearDisplay  \
0                   Life expectancy at age 60 (years)       2000.0   
1     Healthy life expectancy (HALE) at birth (years)       2013.0   
2                   Life expectancy at age 60 (years)       2012.0   
3                   Life expectancy at age 60 (years)       2012.0   
4     Healthy life expectancy (HALE) at birth (years)       2012.0   
...                                               ...          ...   
5240  Healthy life expectancy (HALE) at birth (years)       2000.0   
5241                 Life expectancy at birth (years)       2012.0   
5242  Healthy life expectancy (HALE) at birth (years)       2013.0   
5243                 Life expectancy at birth (years)       1990.0   
5244                 Life expectancy at birth (years)       2000.0   

              RegionDisplay WorldBankIncomeGroupDisplay  SexDisplay  Numeric  
0     Eastern Mediterranean                  Low_income      Female     16.0  
1     Eastern Mediterranean                  Low_income      Female     49.5  
2     Eastern Mediterranean                  Low_income      Female     17.0  
3     Eastern Mediterranean                  Low_income        Male     15.0  
4     Eastern Mediterranean                  Low_income      Female     48.9  
...                     ...                         ...         ...      ...  
5240                 Africa                  Low_income        Male     37.0  
5241                 Africa                  Low_income  Both sexes     58.0  
5242                 Africa                  Low_income      Female     51.8  
5243                 Africa                  Low_income        Male     60.0  
5244                 Africa                  Low_income  Both sexes     42.0  

[5245 rows x 11 columns]

In [55]:
{
    "Строки в drinks": len(drinks),
    "Строки в life_expectancy": len(life_expect),
    "Строки в объединенном датасете": len(df_mrg)
}

{'Строки в drinks': 193,
 'Строки в life_expectancy': 5244,
 'Строки в объединенном датасете': 5245}

### 12

Осуществите группировку данных по странам и типу научной статьи (многомерная группировка). В качестве столбцов значений для группировки возьмите оцененную продолжительность жизни и общее потребление алкоголя в литрах. В качестве агреггирующих функций для продолжительности жизни возьмите среднее, медиану, минимум и максимум, а для потребления алкоголя — агреггирующую функцию/функции выберите самостоятельно. Как вы можете прокомментировать свой выбор и что вы наблюдаете?

In [56]:
grouped_df = df_mrg.groupby(['CountryDisplay', 'GhoDisplay']).agg(
    life_expectency_mean = ('Numeric','mean'),
    life_expectency_meadian = ('Numeric','median'),
    life_expectency_min = ('Numeric','min'),
    life_expectency_max = ('Numeric','max'),
    alco_sum = ('total_litres_of_pure_alcohol', 'sum'),
    alco_mean = ('total_litres_of_pure_alcohol', 'mean')).reset_index()
grouped_df.head(20)

CountryDisplay                                       GhoDisplay  \
0         Afghanistan  Healthy life expectancy (HALE) at birth (years)   
1         Afghanistan                Life expectancy at age 60 (years)   
2         Afghanistan                 Life expectancy at birth (years)   
3             Albania  Healthy life expectancy (HALE) at birth (years)   
4             Albania                Life expectancy at age 60 (years)   
5             Albania                 Life expectancy at birth (years)   
6             Algeria  Healthy life expectancy (HALE) at birth (years)   
7             Algeria                Life expectancy at age 60 (years)   
8             Algeria                 Life expectancy at birth (years)   
9             Andorra  Healthy life expectancy (HALE) at birth (years)   
10            Andorra                Life expectancy at age 60 (years)   
11            Andorra                 Life expectancy at birth (years)   
12             Angola  Healthy life expectancy (HALE) at birth (years)   
13             Angola                Life expectancy at age 60 (years)   
14             Angola                 Life expectancy at birth (years)   
15  Antigua & Barbuda  Healthy life expectancy (HALE) at birth (years)   
16  Antigua & Barbuda                Life expectancy at age 60 (years)   
17  Antigua & Barbuda                 Life expectancy at birth (years)   
18          Argentina  Healthy life expectancy (HALE) at birth (years)   
19          Argentina                Life expectancy at age 60 (years)   

    life_expectency_mean  life_expectency_meadian  life_expectency_min  \
0              47.833333                     48.8                 44.6   
1              15.000000                     15.0                 13.0   
2              54.666667                     55.0                 49.0   
3              63.800000                     63.9                 59.9   
4              17.333333                     18.0                 15.0   
5              71.111111                     71.0                 67.0   
6              61.622222                     61.7                 59.4   
7              17.555556                     18.0                 16.0   
8              69.666667                     70.0                 66.0   
9              70.966667                     71.7                 66.9   
10             23.444444                     23.0                 19.0   
11             79.888889                     80.0                 74.0   
12             42.444444                     43.3                 37.9   
13             14.666667                     15.0                 13.0   
14             46.555556                     46.0                 41.0   
15             64.066667                     64.1                 62.5   
16             19.777778                     20.0                 17.0   
17             73.000000                     73.0                 70.0   
18             66.077778                     66.5                 62.2   
19             20.555556                     21.0                 17.0   

    life_expectency_max  alco_sum  alco_mean  
0                  50.5       0.0        0.0  
1                  17.0       0.0        0.0  
2                  61.0       0.0        0.0  
3                  66.0      44.1        4.9  
4                  20.0      44.1        4.9  
5                  75.0      44.1        4.9  
6                  63.2       6.3        0.7  
7                  19.0       6.3        0.7  
8                  73.0       6.3        0.7  
9                  74.0     111.6       12.4  
10                 28.0     111.6       12.4  
11                 86.0     111.6       12.4  
12                 45.6      53.1        5.9  
13                 16.0      53.1        5.9  
14                 52.0      53.1        5.9  
15                 65.8      44.1        4.9  
16                 23.0      44.1        4.9  
17                 77.0      44.1        4.9  
18  

Касаемо статистики по алкоголю выбрал суммарное значение и среднее, тк обе метрики сравнимы между странами. Например имее среднее одной страны и сравнивая его с средним другой, можно допустить какие-либо полезные выводы, это не работало бы с мин. или макс. тк в целом в таких метриках могут быть выбросы и сравнивать такие значения будет грубо.


Если анализировать получившейся датафрейм, можем заметить, что статистика по алкоголю дублируется в каждой стране на каждый вид статьи. То есть сравнивать по этим значениям разные страны относительной видов статей неккоректно. Говоря о статистики по длине жизни, то тут все хорошо, явно можно проводить сравнения.

### 13

Вставьте любые два столбца из предыдущего задания, которые вы получили путем группировки, в исходную таблицу в качестве двух новых столбцов (назвать вы их можете как хотите). Разумеется, значения этих столбцов должны соответствовать значениям исходной таблицы по указанным в прошлом задании категориям.

*Примечание. Не забудьте сохранить все изменения! В последующих заданиях вы должны работать с обновленным датасетом.*

In [57]:
df_mrg = pd.merge(
    df_mrg,
    grouped_df[['CountryDisplay', 'GhoDisplay', 'life_expectency_mean', 'alco_mean']],
    on=['CountryDisplay', 'GhoDisplay'], how='left')
df_mrg

CountryDisplay  beer_servings  spirit_servings  wine_servings  \
0       Afghanistan            0.0              0.0            0.0   
1       Afghanistan            0.0              0.0            0.0   
2       Afghanistan            0.0              0.0            0.0   
3       Afghanistan            0.0              0.0            0.0   
4       Afghanistan            0.0              0.0            0.0   
...             ...            ...              ...            ...   
5240       Zimbabwe           64.0             18.0            4.0   
5241       Zimbabwe           64.0             18.0            4.0   
5242       Zimbabwe           64.0             18.0            4.0   
5243       Zimbabwe           64.0             18.0            4.0   
5244       Zimbabwe           64.0             18.0            4.0   

      total_litres_of_pure_alcohol  \
0                              0.0   
1                              0.0   
2                              0.0   
3                              0.0   
4                              0.0   
...                            ...   
5240                           4.7   
5241                           4.7   
5242                           4.7   
5243                           4.7   
5244                           4.7   

                                           GhoDisplay  YearDisplay  \
0                   Life expectancy at age 60 (years)       2000.0   
1     Healthy life expectancy (HALE) at birth (years)       2013.0   
2                   Life expectancy at age 60 (years)       2012.0   
3                   Life expectancy at age 60 (years)       2012.0   
4     Healthy life expectancy (HALE) at birth (years)       2012.0   
...                                               ...          ...   
5240  Healthy life expectancy (HALE) at birth (years)       2000.0   
5241                 Life expectancy at birth (years)       2012.0   
5242  Healthy life expectancy (HALE) at birth (years)       2013.0   
5243                 Life expectancy at birth (years)       1990.0   
5244                 Life expectancy at birth (years)       2000.0   

              RegionDisplay WorldBankIncomeGroupDisplay  SexDisplay  Numeric  \
0     Eastern Mediterranean                  Low_income      Female     16.0   
1     Eastern Mediterranean                  Low_income      Female     49.5   
2     Eastern Mediterranean                  Low_income      Female     17.0   
3     Eastern Mediterranean                  Low_income        Male     15.0   
4     Eastern Mediterranean                  Low_income      Female     48.9   
...                     ...                         ...         ...      ...   
5240                 Africa                  Low_income        Male     37.0   
5241                 Africa                  Low_income  Both sexes     58.0   
5242                 Africa                  Low_income      Female     51.8   
5243                 Africa                  Low_income        Male     60.0   
5244                 Africa                  Low_income  Both sexes     42.0   

      life_expectency_mean  alco_mean  
0                15.000000        0.0  
1                47.833333        0.0  
2                15.000000        0.0  
3                15.000000        0.0  
4                47.833333        0.0  
...                    ...        ...  
5240             45.222222        4.7  
5241             54.000000        4.7  
5242             45.222222        4.7  
5243             54.000000        4.7  
5244             54.000000        4.7  

[5245 rows x 13 columns]

### 14

Используя группировки, аргументированно ответьте на вопрос: на каких континентах средняя ожидаемая продолжительность жизни выше у мужчин, а на каких у женщин?

А медианная ожидаемая продолжительность?

А минимальная ожидаемая продолжительность?

Сделайте выводы!

*Подсказка. Казалось бы, при чем здесь тип научной статьи?*

In [58]:
df_at_birth = df_mrg[df_mrg['GhoDisplay'] == 'Life expectancy at birth (years)']
gp_sexes_on_continents = df_at_birth.groupby(['RegionDisplay', 'SexDisplay']).agg(
    life_expectency_mean=('Numeric', 'mean'),
    life_expectency_median=('Numeric', 'median'),
    life_expectency_min=('Numeric', 'min')).reset_index()

gp_sexes_on_continents

RegionDisplay  SexDisplay  life_expectency_mean  \
0                  Africa  Both sexes             54.978261   
1                  Africa      Female             56.594203   
2                  Africa        Male             53.463768   
3                Americas  Both sexes             72.419048   
4                Americas      Female             75.257143   
5                Americas        Male             69.695238   
6   Eastern Mediterranean  Both sexes             67.060606   
7   Eastern Mediterranean      Female             68.818182   
8   Eastern Mediterranean        Male             65.515152   
9                  Europe  Both sexes             74.572327   
10                 Europe      Female             77.861635   
11                 Europe        Male             71.251572   
12        South_East Asia  Both sexes             64.969697   
13        South_East Asia      Female             66.757576   
14        South_East Asia        Male             63.151515   
15        Western Pacific  Both sexes             70.135802   
16        Western Pacific      Female             72.728395   
17        Western Pacific        Male             67.703704   

    life_expectency_median  life_expectency_min  
0                     54.5                 38.0  
1                     56.0                 38.0  
2                     53.5                 38.0  
3                     73.0                 54.0  
4                     76.0                 56.0  
5                     70.0                 52.0  
6                     70.0                 42.0  
7                     71.0                 44.0  
8                     67.5                 41.0  
9                     75.0                 62.0  
10                    79.0                 65.0  
11                    72.0                 58.0  
12                    66.0                 50.0  
13                    68.0                 51.0  
14                    65.0                 48.0  
15                    70.0                 53.0  
16                    73.0                 54.0  
17                    68.0                 51.0

Для реализации этого задания нам необходимо выбрать записи с типом научной статьи, которая отражает ожидаемую длину жизни от рождения. Если игнорировать это, то мы испортим статистику неподходящими данными. Например, если бы мы оставили тип научной статьи ожидаемая длина жизни после 60, то очевидно бы пришли с абсурдным результатам, по типу мин. длина жизни 5 лет, средняя 15 и тд.

Во всех континентах средняя продолжительность жизни выше у женщин. Это же касается медианной продолжительности жизни, у женщин она выше. Говоря о мин. продолжительности жизни, то есть небольшие новшества. Основой тренд сохранен (у женщин мин. продолжительность выше), но в регионах Афирка и Юго_восточная Азия эта метрика у полов совпадает.

#**Вывод:**

Женщины живут дольше, чем мужчины, это правило расспростроняется на все континенты.

### 15

Работая лишь с общим объединенным датасетом и не используя исходную таблицу drinks, подсчитайте среднее потребление каждого вида алкоголя и среднее общее количество литров потребляемого алкоголя на человека по всем континентам. Занесите ваши результаты в отдельный датафрейм!

*Подсказка. Задача немного сложнее, чем может показаться на первый взгляд — в объединенном датасете страны встречаются разное количество раз, и из-за этого могут возникнуть определенные проблемы. Если так случится, надо подумать, как это обойти!*

*Примечание. Разумеется, это необходимо сделать автоматизированно, а не перебирая каждый континент по отдельности.*

In [59]:
df_drinks_stat = df_mrg[['CountryDisplay', 'RegionDisplay', 'beer_servings',
'spirit_servings', 'wine_servings', 'total_litres_of_pure_alcohol']].drop_duplicates()

df_drinks_stat = df_drinks_stat.groupby('RegionDisplay').agg(
    avg_beer=('beer_servings', 'mean'),
    avg_spirit=('spirit_servings', 'mean'),
    avg_wine=('wine_servings', 'mean'),
    avg_total=('total_litres_of_pure_alcohol', 'mean')).reset_index()

df_drinks_stat

RegionDisplay    avg_beer  avg_spirit    avg_wine  avg_total
0                 Africa   69.733333   17.000000   17.933333   3.380000
1               Americas  155.600000  148.257143   37.514286   6.102857
2  Eastern Mediterranean   10.428571   21.190476    4.857143   0.714286
3                 Europe  175.132075  130.415094  123.528302   7.873585
4        South_East Asia   14.818182   44.090909    0.454545   1.063636
5        Western Pacific   81.962963   66.851852   27.740741   3.637037

Чтобы решить проблему повторной встречи одной и тоей же страны несколько раз я удалил дубликаты по нескольким признакам, потому что, как намекнул автор задания, они могут создавать проблемы - скорее всего имелись в виду перевесы.

### 16

Создание новых признаков (feature engineering) является одним из основных средств улучшения качества работы алгоритмов машинного обучения на этапе обработки данных, а также отличный способ увеличить глубину ваших выводов из проделанного анализа в целом! Частично мы с вами уже осуществляли feature engineering, когда добавляли новые столбцы в задании 13 данного раздела.

Теперь давайте добавим в таблицу столбец, который будет показывать, какое общее количество servings per person производится в среднем по стране, к которой относится научная статья. Столбец для этого признака назовите 'total_servings'.

Чем может быть полезен такой признак?

***Бонус (2 бон. балла):*** Придумайте еще какой-нибудь свой осмысленный признак, который может быть полезен для анализа (несет в себе какую-то информацию), и добавьте его в таблицу. Обязательно аргументируйте, чем полезен ваш признак.

In [60]:
total_servings = df_mrg.groupby('CountryDisplay')[['beer_servings',
'spirit_servings', 'wine_servings']].mean().sum(axis=1)

df_mrg['total_servings'] = df_mrg['CountryDisplay'].map(total_servings)

df_mrg

CountryDisplay  beer_servings  spirit_servings  wine_servings  \
0       Afghanistan            0.0              0.0            0.0   
1       Afghanistan            0.0              0.0            0.0   
2       Afghanistan            0.0              0.0            0.0   
3       Afghanistan            0.0              0.0            0.0   
4       Afghanistan            0.0              0.0            0.0   
...             ...            ...              ...            ...   
5240       Zimbabwe           64.0             18.0            4.0   
5241       Zimbabwe           64.0             18.0            4.0   
5242       Zimbabwe           64.0             18.0            4.0   
5243       Zimbabwe           64.0             18.0            4.0   
5244       Zimbabwe           64.0             18.0            4.0   

      total_litres_of_pure_alcohol  \
0                              0.0   
1                              0.0   
2                              0.0   
3                              0.0   
4                              0.0   
...                            ...   
5240                           4.7   
5241                           4.7   
5242                           4.7   
5243                           4.7   
5244                           4.7   

                                           GhoDisplay  YearDisplay  \
0                   Life expectancy at age 60 (years)       2000.0   
1     Healthy life expectancy (HALE) at birth (years)       2013.0   
2                   Life expectancy at age 60 (years)       2012.0   
3                   Life expectancy at age 60 (years)       2012.0   
4     Healthy life expectancy (HALE) at birth (years)       2012.0   
...                                               ...          ...   
5240  Healthy life expectancy (HALE) at birth (years)       2000.0   
5241                 Life expectancy at birth (years)       2012.0   
5242  Healthy life expectancy (HALE) at birth (years)       2013.0   
5243                 Life expectancy at birth (years)       1990.0   
5244                 Life expectancy at birth (years)       2000.0   

              RegionDisplay WorldBankIncomeGroupDisplay  SexDisplay  Numeric  \
0     Eastern Mediterranean                  Low_income      Female     16.0   
1     Eastern Mediterranean                  Low_income      Female     49.5   
2     Eastern Mediterranean                  Low_income      Female     17.0   
3     Eastern Mediterranean                  Low_income        Male     15.0   
4     Eastern Mediterranean                  Low_income      Female     48.9   
...                     ...                         ...         ...      ...   
5240                 Africa                  Low_income        Male     37.0   
5241                 Africa                  Low_income  Both sexes     58.0   
5242                 Africa                  Low_income      Female     51.8   
5243                 Africa                  Low_income        Male     60.0   
5244                 Africa                  Low_income  Both sexes     42.0   

      life_expectency_mean  alco_mean  total_servings  
0                15.000000        0.0             0.0  
1                47.833333        0.0             0.0  
2                15.000000        0.0             0.0  
3                15.000000        0.0             0.0  
4                47.833333        0.0             0.0  
...                    ...        ...             ...  
5240             45.222222        4.7            86.0  
5241             54.000000        4.7            86.0  
5242             45.222222        4.7            86.0  
5243             54.000000        4.7            86.0  
5244             54.000000        4.7            86.0  

[5245 rows x 14 columns]

Признак полезен для упрощенного сравнения стран, ведь мы теперь у нас имеется 1 четкая метрика, а не 3 признака.

In [61]:
df_mrg['pure_alcohol_per_serving'] = df_mrg['total_litres_of_pure_alcohol']/df_mrg['total_servings']
df_mrg

CountryDisplay  beer_servings  spirit_servings  wine_servings  \
0       Afghanistan            0.0              0.0            0.0   
1       Afghanistan            0.0              0.0            0.0   
2       Afghanistan            0.0              0.0            0.0   
3       Afghanistan            0.0              0.0            0.0   
4       Afghanistan            0.0              0.0            0.0   
...             ...            ...              ...            ...   
5240       Zimbabwe           64.0             18.0            4.0   
5241       Zimbabwe           64.0             18.0            4.0   
5242       Zimbabwe           64.0             18.0            4.0   
5243       Zimbabwe           64.0             18.0            4.0   
5244       Zimbabwe           64.0             18.0            4.0   

      total_litres_of_pure_alcohol  \
0                              0.0   
1                              0.0   
2                              0.0   
3                              0.0   
4                              0.0   
...                            ...   
5240                           4.7   
5241                           4.7   
5242                           4.7   
5243                           4.7   
5244                           4.7   

                                           GhoDisplay  YearDisplay  \
0                   Life expectancy at age 60 (years)       2000.0   
1     Healthy life expectancy (HALE) at birth (years)       2013.0   
2                   Life expectancy at age 60 (years)       2012.0   
3                   Life expectancy at age 60 (years)       2012.0   
4     Healthy life expectancy (HALE) at birth (years)       2012.0   
...                                               ...          ...   
5240  Healthy life expectancy (HALE) at birth (years)       2000.0   
5241                 Life expectancy at birth (years)       2012.0   
5242  Healthy life expectancy (HALE) at birth (years)       2013.0   
5243                 Life expectancy at birth (years)       1990.0   
5244                 Life expectancy at birth (years)       2000.0   

              RegionDisplay WorldBankIncomeGroupDisplay  SexDisplay  Numeric  \
0     Eastern Mediterranean                  Low_income      Female     16.0   
1     Eastern Mediterranean                  Low_income      Female     49.5   
2     Eastern Mediterranean                  Low_income      Female     17.0   
3     Eastern Mediterranean                  Low_income        Male     15.0   
4     Eastern Mediterranean                  Low_income      Female     48.9   
...                     ...                         ...         ...      ...   
5240                 Africa                  Low_income        Male     37.0   
5241                 Africa                  Low_income  Both sexes     58.0   
5242                 Africa                  Low_income      Female     51.8   
5243                 Africa                  Low_income        Male     60.0   
5244                 Africa                  Low_income  Both sexes     42.0   

      life_expectency_mean  alco_mean  total_servings  \
0                15.000000        0.0             0.0   
1                47.833333        0.0             0.0   
2                15.000000        0.0             0.0   
3                15.000000        0.0             0.0   
4                47.833333        0.0             0.0   
...                    ...        ...             ...   
5240             45.222222        4.7            86.0   
5241             54.000000        4.7            86.0   
5242             45.222222        4.7            86.0   
5243             54.000000        4.7            86.0   
5244             54.000000        4.7            86.0   

      pure_alcohol_per_serving  
0                          NaN  
1                          NaN  
2                          NaN  
3                          NaN  
4                          NaN  
...                   

Я решил добавить признак pure_alcohol_per_serving, который показывает среднее количество литров чистого алко. на порцию. Таким образом мы можем оценить, насколько крепкое алко. в среднем употребляют в каждой стране.

### 17

Есть ли в данных категориальные признаки, которые принимают значения строк, а не чисел? Если да, то для многих алгоритмов это может стать серьезным препятствием, поэтому закодируйте категориальные признаки так, чтобы к данным можно было применить любые алгоритмы. Аргументируйте выбор методов кодирования.

*Подсказка. Вам могут помочь Label encoding и One-Hot Encoding. Но будьте внимательны! У каждого из этих методов есть свои сильные и слабые стороны, а также ограничения применимости.*

In [62]:
from sklearn.preprocessing import LabelEncoder

one_hot_columns = ['GhoDisplay', 'RegionDisplay', 'CountryDisplay', 'SexDisplay']
df_mrg = pd.get_dummies(df_mrg, columns=one_hot_columns, drop_first=True)

label_encoder = LabelEncoder()
df_mrg['WorldBankIncomeGroupDisplay'] = label_encoder.fit_transform(df_mrg['WorldBankIncomeGroupDisplay'])

df_mrg.head()

beer_servings  spirit_servings  wine_servings  \
0            0.0              0.0            0.0   
1            0.0              0.0            0.0   
2            0.0              0.0            0.0   
3            0.0              0.0            0.0   
4            0.0              0.0            0.0   

   total_litres_of_pure_alcohol  YearDisplay  WorldBankIncomeGroupDisplay  \
0                           0.0       2000.0                            1   
1                           0.0       2013.0                            1   
2                           0.0       2012.0                            1   
3                           0.0       2012.0                            1   
4                           0.0       2012.0                            1   

   Numeric  life_expectency_mean  alco_mean  total_servings  ...  \
0     16.0             15.000000        0.0             0.0  ...   
1     49.5             47.833333        0.0             0.0  ...   
2     17.0             15.000000        0.0             0.0  ...   
3     15.0             15.000000        0.0             0.0  ...   
4     48.9             47.833333        0.0             0.0  ...   

   CountryDisplay_Uruguay  CountryDisplay_Uzbekistan  CountryDisplay_Vanuatu  \
0                   False                      False                   False   
1                   False                      False                   False   
2                   False                      False                   False   
3                   False                      False                   False   
4                   False                      False                   False   

   CountryDisplay_Venezuela  CountryDisplay_Vietnam  CountryDisplay_Yemen  \
0                     False                   False                 False   
1                     False                   False                 False   
2                     False                   False                 False   
3                     False                   False                 False   
4                     False                   False                 False   

   CountryDisplay_Zambia  CountryDisplay_Zimbabwe  SexDisplay_Female  \
0                  False                    False               True   
1                  False                    False               True   
2                  False                    False               True   
3                  False                    False              False   
4                  False                    False               True   

   SexDisplay_Male  
0            False  
1            False  
2            False  
3             True  
4            False  

[5 rows x 214 columns]

Прочитав плюсы и минусы обоих видов кодировкок пришла идея поступить след. образом - признаки не имеющие ранжирования (напр. название стран, вид статьи) я закодировал one-hot encoding, дабы не создать иллюзию порядка. Но для колонки WorldBankIncomeGroupDisplay был использован label encoding, тк это признак имеющий порядок (малый, средний, высокий уровень). Таким образом мы избавились от лишних 2 колонок, что сэкономит память и придаст удобства, так и еще потенциально улучшится точность моделей, основаных на числах, хороший пример линейная регрессия.

### 18

Придумайте какое-нибудь своё интересное задание на анализ предлагаемых данных и решите его.

Креативность в этом задании приветствуется, и за нее можно получить ***1-2 бонусных балла***!

*Подсказка. Вы можете вдохновляться различными заданиями из этого раздела ДЗ, как, например, заданиями: 4, 6, 10, 12, 14, 15, 16. Но, чтобы получить бонусные баллы, лучше не используйте аналогичные идеи, а придумайте что-то своё!*

In [63]:
at_bitrh_both_sexes = life_expect[
    (life_expect['GhoDisplay'] == 'Life expectancy at birth (years)') &
(life_expect['SexDisplay'] == 'Both sexes')]

at_bitrh_both_sexes = at_bitrh_both_sexes.groupby('RegionDisplay').agg(
    first_year=('YearDisplay', 'min'),
    last_year=('YearDisplay', 'max'),
    start_life_expectancy=('Numeric', 'first'),
    end_life_expectancy=('Numeric', 'last'))

at_bitrh_both_sexes['change'] = (at_bitrh_both_sexes['end_life_expectancy'] -
at_bitrh_both_sexes['start_life_expectancy'])

at_bitrh_both_sexes = at_bitrh_both_sexes.sort_values(by='change', ascending=False)
at_bitrh_both_sexes

first_year  last_year  start_life_expectancy  \
RegionDisplay                                                         
South_East Asia              1990       2012                   59.0   
Eastern Mediterranean        1990       2012                   61.0   
Africa                       1990       2012                   50.0   
Americas                     1990       2012                   80.0   
Europe                       1990       2012                   70.0   
Western Pacific              1990       2012                   79.0   

                       end_life_expectancy  change  
RegionDisplay                                       
South_East Asia                       68.0     9.0  
Eastern Mediterranean                 68.0     7.0  
Africa                                51.0     1.0  
Americas                              79.0    -1.0  
Europe                                68.0    -2.0  
Western Pacific                       68.0   -11.0

**Задача:**
Создать новый датасет отражающий изменения ожидаемой продолжительности жизни с рождения на разных континентах за период с 1990 по 2012 годы, определить регионы с наибольшим прогрессом и регрессом.

**Решение:**

Для начала я решил создать отдельный отфильтрованный по типу статьи и виду пола датасет. Далее данные были сагрегированы по континентам, вывели начальный год и конечный, значение ожидаемое длительной жизни в начальном году и конечном. Далее создал столбец change, который выражает разность между ожидаемой длительностью на год окончания и начала исследования.
В конце добавлена сортировка по столбцу change.


**Результаты:**


Видим позитивную динамику у Юго-Восточной Азии (+9 лет!), Восточное Средиземноморье +7 лет, Африка +1 год. Америка -1 год, в пределах погрешности, сохранила уровень. У Европы произошло снижение на 2 года. У Восточной части Тихого Океана не все так радужно - целых -11 лет.



## Бонусная часть на анализ целевой переменной (до 5 бонусных баллов)

После всего проделанного анализа и подготовки данных мы можем перейти непосредственно к ответу на исходный вопрос исследования, который был сформулирован вами в задании 3 данного раздела.

### 19

***Бонус (5 бон. баллов):*** Дайте развернутый, обоснованный, подкрепленный данными и написанным вами кодом ответ на вопрос, влияют ли признаки зависимой таблицы на целевую переменную из главной таблицы и, если да, то каким именно образом влияют?

Обязательно аргументируйте полученные вами результаты и выводы. Чем подробнее и обоснованнее анализ вы проведете в рамках данного бонусного задания, тем больше бонусных баллов получите! Вы можете использовать подходы и методы из рассмотренных на семинаре и в рамках данного домашнего задания, а можете придумать и что-то своё. Креативность максимально приветствуется!

In [ ]:
# your code here

    {your final conclusion here}